tests/Application/Actions/Donations/UpdateTest.php

```diff
@@ -1218,6 +1219,210 @@ public function testAddDataHitsAlreadyCapturedStripeExceptionWithFeeChange(): vo
         ], $payloadArray);
     }
 
+    /**
+     * Success on retry.
+     */
+    public function testAddDataHitsStripeLockExceptionOnce(): void
+    {
+        $app = $this->getAppInstance();
+        /** @var Container $container */
+        $container = $app->getContainer();
+
+        $donation = $this->getTestDonation();
+        $donation->setDonorCountryCode('US');
+        $donation->setCurrencyCode('USD');
+        $donation->setTipAmount('3.21');
+        $donation->setGiftAid(true);
+        $donation->setTipGiftAid(false);
+        $donation->setDonorHomeAddressLine1('99 Updated St');
+        $donation->setDonorHomePostcode('X1 1XY');
+        $donation->setDonorFirstName('Saul');
+        $donation->setDonorLastName('Williams');
+        $donation->setDonorEmailAddress('saul@example.com');
+        $donation->setTbgComms(true);
+        $donation->setCharityComms(false);
+        $donation->setChampionComms(false);
+        $donation->setDonorBillingAddress('Y1 1YX');
+
+        $donationRepoProphecy = $this->prophesize(DonationRepository::class);
+        $donationRepoProphecy
+            ->findOneBy(['uuid' => '12345678-1234-1234-1234-1234567890ab'])
+            ->willReturn($this->getTestDonation()) // Get a new mock object so DB has old values.
+            ->shouldBeCalledOnce();
+        $donationRepoProphecy
+            ->push(Argument::type(Donation::class), false)
+            ->shouldBeCalledOnce();
+        $donationRepoProphecy
+            ->deriveFees(Argument::type(Donation::class), null, null)
+            ->willReturn($donation) // Actual fee calculation is tested elsewhere.
+            ->shouldBeCalledOnce();
+
+        // Persist as normal.
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $entityManagerProphecy->beginTransaction()->shouldBeCalledOnce();
+        $entityManagerProphecy->persist(Argument::type(Donation::class))->shouldBeCalledOnce();
+        $entityManagerProphecy->flush()->shouldBeCalledOnce();
+        $entityManagerProphecy->commit()->shouldBeCalledOnce();
```

Comment:
Why are we mocking out both the repositories and the entity manager? It seems a bit confusing to be mocking out two layers at once. I guess there must be some parts of the code under test using the repositories and other parts going straight through to the EM.

Reply:
@bdsl Yes, that's why. You can see in the corresponding Action's constructor where this happens, and I think its (`Update`'s) code shows why it's quite a helpful choice.